### PR TITLE
Debian 12 UEFI provisioning

### PR DESCRIPTION
**Pull request by stejskalleos**

### CODE REVIEW

File: boot/debian_12/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+cd ./boot/debian_12
+
+# Kernel
+wget -O vmlinuz http://http.us.debian.org/debian/dists/bookworm/main/installer-amd64/current/images/netboot/debian-installer/amd64/linux
+
+# InitRAM disk
+wget http://http.us.debian.org/debian/dists/bookworm/main/installer-amd64/current/images/netboot/debian-installer/amd64/initrd.gz
+
+# Grub loader
+wget http://http.us.debian.org/debian/dists/bookworm/main/installer-amd64/current/images/netboot/debian-installer/amd64/bootnetx64.efi -O grubx64.efi
```
Comment (stejskalleos) [inline boot/debian_12/README.md:17]: It's failing on:
```
python3 -m http.server --protocol HTTP/1.1
192.168.122.16 - - [26/Feb/2024 13:59:49] "HEAD /grubx64.efi HTTP/1.1" 200 -
192.168.122.16 - - [26/Feb/2024 13:59:49] "GET /grubx64.efi HTTP/1.1" 200 -
----------------------------------------
Exception occurred during processing of request from ('192.168.122.16', 1863)
Traceback (most recent call last):
  File "/usr/lib64/python3.12/socketserver.py", line 692, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/lib64/python3.12/http/server.py", line 1311, in finish_request
    self.RequestHandlerClass(request, client_address, self,
  File "/usr/lib64/python3.12/http/server.py", line 672, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/lib64/python3.12/socketserver.py", line 761, in __init__
    self.handle()
  File "/usr/lib64/python3.12/http/server.py", line 438, in handle
    self.handle_one_request()
  File "/usr/lib64/python3.12/http/server.py", line 404, in handle_one_request
    self.raw_requestline = self.rfile.readline(65537)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/socket.py", line 707, in readinto
    return self._sock.recv_into(b)
           ^^^^^^^^^^^^^^^^^^^^^^^
ConnectionResetError: [Errno 104] Connection reset by peer
----------------------------------------
192.168.122.16 - - [26/Feb/2024 13:59:49] "GET //grubx64.efi HTTP/1.1" 200 -
----------------------------------------
Exception occurred during processing of request from ('192.168.122.16', 1864)
Traceback (most recent call last):
  File "/usr/lib64/python3.12/socketserver.py", line 692, in process_request_thread
    self.finish_request(request, client_address)

```

Doesn't matter if I run it under apache or just simple python server

Comment (stejskalleos) [reply boot/debian_12/README.md:17]: Not sure why it's calling `GET //grubx64.efi`

File: README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+dnf install @virtualization dnf-plugins-core
+dnf config-manager --add-repo http://www.kraxel.org/repos/firmware.repo
+dnf install edk2.git-ovmf-x64
+```
+
+Enable Libvirt
+
+```console
+systemctl start libvirtd
+systemctl enable libvirtd
```
Comment (ekohl) [inline README.md:19]: You can do both in a single command. And you indicate it's console, so you can use `#` (root) or `$` (non-root) to indicate it's an actual command.
```suggestion
# systemctl enable --now libvirtd
```

File: README.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+```console
+virsh net-destroy default; virsh net-start default
+```
+
+## HTTP server
+
+Run the HTTP server FROM the OS directory
+
+```console
+python3 -m http.server --protocol HTTP/1.1
```
Comment (ekohl) [inline README.md:41]: I'd add a note you need at least Python 3.11+ for the `--protocol` parameter.

File: boot/centos_9/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# CentOS 9 provisioning
+
+```console
```
Comment (ekohl) [inline boot/centos_9/README.md:3]: This looks more like `sh` and less like `console`

File: boot/centos_9/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+Getting Shim
+
+- Spin up CentOS 9 Stream machine
+- Update all the packages
+- Get the Shim from the machine
+
+```console
+
+```
```
Comment (ekohl) [inline boot/centos_9/README.md:26]: Empty block?

File: README.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+(Tested on Fedora 39)
+
+## Preconditions
+
+Required packages
+
+```console
+dnf install @virtualization dnf-plugins-core
+dnf config-manager --add-repo http://www.kraxel.org/repos/firmware.repo
+dnf install edk2.git-ovmf-x64
```
Comment (ekohl) [inline README.md:12]: I think in recent Fedoras you don't need that repo and can use `edk2-ovmf`
```suggestion
dnf install edk2-ovmf-x64
```

File: README.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+systemctl start libvirtd
+systemctl enable libvirtd
+```
+
+## Libvirt network
+
+Configure network same way as in `data/libvirt_network.xml`
+
+```console
+virsh net-edit default
```
Comment (ekohl) [inline README.md:27]: You can directly create a network from a file, so I think this should work:
```suggestion
virsh net-create data/libvirt_network.xml
```

File: boot/debian_12/README.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+# This one doesn't work
+#wget http://http.us.debian.org/debian/dists/bookworm/main/installer-amd64/current/images/netboot/debian-installer/amd64/grubx64.efi
+
+sudo chown -R $USER: .
```
Comment (ekohl) [inline boot/debian_12/README.md:22]: chown shouldn't be needed since you're downloading them.

File: boot/debian_12/README.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
+# Debian 12
+
+- [Repo](http://http.us.debian.org/debian/dists/bookworm/main/installer-amd64/current/images/netboot/debian-installer/amd64)
+
+## Preparations
+
+```console
```
Comment (ekohl) [inline boot/debian_12/README.md:7]: Can I suggest saving this as `boot/debian_12/Makefile`
```make
baseurl = https://deb.debian.org/debian/dists/bookworm/main/installer-amd64/current/images/netboot/debian-installer/amd64

all: vmlinuz initrd.gz grubx64.efi

vmlinuz:
	wget -O $@ $(baseurl)/linux

initrd.gz:
	wget -O $@ $(baseurl)/initrd.gz

grubx64.efi:
	wget -O $@ $(baseurl)/bootnetx64.efi
```

Then the instructions become:
```sh
make -C boot/debian_12
```

And then apply the same to CentOS 9.

**Review by ekohl — COMMENTED** [review]
My initial notes from reading this. Going to try it out myself now.

**Comment by ekohl** [comment]
This is what I have on my own server in `/etc/dhcp/dhcpd6.conf`:
```
#
# DHCPv6 Server Configuration file.
#   see /usr/share/doc/dhcp-server/dhcpd6.conf.example
#   see dhcpd.conf(5) man page
#

# https://documentation.suse.com/sles/15-SP2/html/SLES-all/cha-deployment-prep-uefi-httpboot.html#httpboot-dhcpv6-server
option dhcp6.bootfile-url code 59 = string;
option dhcp6.vendor-class code 16 = {integer 32, integer 16, string};
subnet6 2001:db8::/64 {
        range6 2001:db8:0:0:cafe:: 2001:db8:0:0:cafe:ffff:ffff:ffff;
        option dhcp6.bootfile-url "https://smartproxy.example.com:8443/EFI/grub2/shim.efi";
        option dhcp6.name-servers 2001:db8::1;
        option dhcp6.vendor-class 0 10 "HTTPClient";
}
```

This at least allowed me to provision with it, but I'll admit everything was CentOS 8 Stream.

File: boot/debian_12/README.md (hunk 1 above)
Comment (ekohl) [reply boot/debian_12/README.md:17]: In my testing with starting `bootnetx64.efi` (which chainloads `grubx64.efi`) I came to the conclusion that the grub image is built without the HTTP module, so it doesn't request the `grub.cfg` file. Though I haven't 100% verified it, I didn't see any request to my server. When I tried `insmod http` I did see `No DNS server configured` which leads me to https://bugzilla.redhat.com/show_bug.cgi?id=1842509.

I couldn't get my TFTP environment working for some reason and ran out of time experimenting.

**Pull request closed without merging**